File: impl/sdk/java/src/examples/java/io/resonatehq/examples/pipeline/Pipeline.java
package io.resonatehq.examples.pipeline;

import io.resonatehq.resonate.Context;
import io.resonatehq.resonate.Context.ResonateFuture;
import io.resonatehq.resonate.Handle.ResonateHandle;
import io.resonatehq.resonate.Resonate;
import java.util.List;

/**
 * pipeline shows a multi-stage DAG-shaped durable workflow:
 *
 * <pre>
 *     download -&gt; parse -&gt; +- transform_a -+
 *                          +- transform_b -+- merge -&gt; emit
 * </pre>
 *
 * <p>transform_a and transform_b run in parallel (both calls are dispatched before either is
 * awaited); merge depends on both and synchronizes them with await. Every stage is a registered
 * function backed by a durable promise, so a crash mid-pipeline picks up at the first unsettled
 * stage without re-doing completed work.
 *
 * <p>Note on replay: a durable orchestrator re-executes from the top each time it awaits a
 * not-yet-settled future, so any side effect (a {@code println}, an external call) belongs in a leaf
 * stage function -- which settles once and never re-runs -- not in {@code runPipeline} itself. That
 * is why every log line below lives in a stage.
 */
public final class Pipeline {
    private Pipeline() {}

    /** The pair returned by {@code merge} and unpacked into {@code emit}. */
    public record Merged(int wordCount, String upper) {}

    public static String download(Context ctx, String url) {
        String body = "the quick brown fox jumps over " + url;
        System.out.printf("  [download] %s -> %d bytes%n", url, body.length());
        return body;
    }

    public static List<String> parse(Context ctx, String raw) {
        List<String> words = List.of(raw.split(" "));
        System.out.printf("  [parse] %d words%n", words.size());
        return words;
    }

    public static int transformA(Context ctx, List<String> p) {
        System.out.println("  [transform_a] counting words");
        return p.size();
    }

    public static String transformB(Context ctx, List<String> p) {
        String upper = String.join(" ", p).toUpperCase();
        System.out.printf("  [transform_b] uppercased %d chars%n", upper.length());
        return upper;
    }

    public static Merged merge(Context ctx, int wordCount, String upper) {
        System.out.println("  [merge] combining transforms");
        return new Merged(wordCount, upper);
    }

    public static String emit(Context ctx, int wordCount, String upper) {
        System.out.printf("  [emit] words=%d upper=%s%n", wordCount, upper);
        return "ok";
    }

    public static String runPipeline(Context ctx, String url) {
        String raw = ctx.run(Pipeline::download, url).await();
        List<String> parsed = ctx.run(Pipeline::parse, raw).await();

        // fan out: transform_a and transform_b dispatched before either is awaited
        ResonateFuture<Integer> fa = ctx.run(Pipeline::transformA, parsed);
        ResonateFuture<String> fb = ctx.run(Pipeline::transformB, parsed);

        // fan in
        int a = fa.await();
        String b = fb.await();

        Merged merged = ctx.run(Pipeline::merge, a, b).await();
        return ctx.run(Pipeline::emit, merged.wordCount(), merged.upper()).await();
    }

    public static void main(String[] args) {
        String url = System.getenv().getOrDefault("RESONATE_URL", "http://localhost:8001");
        Resonate r = Resonate.builder().url(url).build();
        r.register(Pipeline::runPipeline);

        try {
            String id = "pipeline-" + System.nanoTime();
            System.out.println("[run_pipeline] starting workflow id=" + id);
            ResonateHandle<String> handle = r.run(id, Pipeline::runPipeline, "example.com/doc");
            String out = handle.result();
            assert "ok".equals(out);
            System.out.println("[run_pipeline] OK: sent=" + out);
        } finally {
            r.stop().join();
        }
    }
}

File: impl/sdk/ts/examples/generator/pipeline.ts
// pipeline shows a multi-stage DAG-shaped durable workflow:
//
//   download → parse → ┬─ transformA ─┐
//                      └─ transformB ─┴─ merge → emit
//
// transformA and transformB run in parallel (both started via beginRpc before
// either is awaited); merge depends on both and synchronizes them. Every stage
// is a registered function backed by a durable promise, so a crash mid-pipeline
// picks up at the first unsettled stage without re-doing completed work.
//
// Start a Resonate server on localhost:8001 first (`resonate dev`), then:
//
//   npx tsx examples/generator/pipeline.ts

import { type Context, Resonate } from "../../src/index.js";

interface Merged {
  wordCount: number;
  upper: string;
}

// -- Stage functions (leaves: each does its work and returns) -------------

async function download(_ctx: Context, url: string): Promise<string> {
  const body = `the quick brown fox jumps over ${url}`;
  console.log(`  [download] ${url} -> ${body.length} bytes`);
  return body;
}

async function parse(_ctx: Context, body: string): Promise<string[]> {
  const words = body.split(/\s+/).filter(Boolean);
  console.log(`  [parse] ${words.length} words`);
  return words;
}

async function transformA(_ctx: Context, words: string[]): Promise<number> {
  console.log("  [transformA] counting words");
  return words.length;
}

async function transformB(_ctx: Context, words: string[]): Promise<string> {
  const upper = words.join(" ").toUpperCase();
  console.log(`  [transformB] uppercased ${upper.length} chars`);
  return upper;
}

async function merge(_ctx: Context, wordCount: number, upper: string): Promise<Merged> {
  console.log("  [merge] combining transforms");
  return { wordCount, upper };
}

async function emit(_ctx: Context, m: Merged): Promise<string> {
  console.log(`  [emit] words=${m.wordCount} upper=${JSON.stringify(m.upper)}`);
  return "ok";
}

// -- Pipeline orchestrator ------------------------------------------------

function* runPipeline(ctx: Context, url: string): Generator<any, string, any> {
  const body: string = yield* ctx.rpc<string>("download", url);
  const words: string[] = yield* ctx.rpc<string[]>("parse", body);

  // fan out: transformA and transformB start in parallel...
  const fA = yield* ctx.beginRpc<number>("transformA", words);
  const fB = yield* ctx.beginRpc<string>("transformB", words);

  // ...fan in: await both
  const wordCount: number = yield* fA;
  const upper: string = yield* fB;

  const merged: Merged = yield* ctx.rpc<Merged>("merge", wordCount, upper);
  return yield* ctx.rpc<string>("emit", merged);
}

const resonate = new Resonate({ url: process.env.RESONATE_URL ?? "http://localhost:8001" });
resonate.register("runPipeline", runPipeline);
resonate.register("download", download);
resonate.register("parse", parse);
resonate.register("transformA", transformA);
resonate.register("transformB", transformB);
resonate.register("merge", merge);
resonate.register("emit", emit);

try {
  const id = `pipeline-${Date.now()}`;
  console.log(`[runPipeline] starting workflow id=${id}`);
  const out = await resonate.run(id, runPipeline, "example.com/doc");
  console.log(`[runPipeline] OK: sent=${out}`);
} finally {
  await resonate.stop();
}
